Firestore/core/test/firebase/firestore/remote/grpc_connection_test.mm

```diff
@@ -0,0 +1,169 @@
+  }
+  void OnStreamRead(const grpc::ByteBuffer& message) override {
+  }
+  void OnStreamFinish(const util::Status& status) override {
+    if (IsConnectivityChange(status)) {
+      ++connectivity_change_count_;
+    }
+  }
+
+  int connectivity_change_count() const {
+    return connectivity_change_count_;
+  }
+  int connectivity_change_count_ = 0;
+};
+
+}  // namespace
+
+class GrpcConnectionTest : public testing::Test {
+ public:
+  GrpcConnectionTest() {
+    auto connectivity_monitor_owning =
+        absl::make_unique<MockConnectivityMonitor>(&tester->worker_queue());
+    connectivity_monitor = connectivity_monitor_owning.get();
+    tester = absl::make_unique<GrpcStreamTester>(
+        std::move(connectivity_monitor_owning));
+  }
+
+  void SetNetworkStatus(NetworkStatus new_status) {
+    tester->worker_queue().EnqueueBlocking(
+        [&] { connectivity_monitor->set_status(new_status); });
+    // Make sure the callback executes.
+    tester->worker_queue().EnqueueBlocking([] {});
+  }
+
+  std::unique_ptr<GrpcStreamTester> tester;
+  MockConnectivityMonitor* connectivity_monitor = nullptr;
+};
+
+TEST_F(GrpcConnectionTest, GrpcStreamsNoticeChangeInConnectivity) {
+  ConnectivityObserver observer;
+
+  auto stream = tester->CreateStream(&observer);
+  stream->Start();
+  EXPECT_EQ(observer.connectivity_change_count(), 0);
+
+  SetNetworkStatus(NetworkStatus::Available);
+  // Same status shouldn't trigger a callback.
+  EXPECT_EQ(observer.connectivity_change_count(), 0);
+
+  tester->KeepPollingGrpcQueue();
+  SetNetworkStatus(NetworkStatus::Unavailable);
+  EXPECT_EQ(observer.connectivity_change_count(), 1);
+}
+
+TEST_F(GrpcConnectionTest, GrpcStreamingCallsNoticeChangeInConnectivity) {
+  int change_count = 0;
+  auto streaming_call = tester->CreateStreamingReader();
+  streaming_call->Start(
+      [&](const StatusOr<std::vector<grpc::ByteBuffer>>& result) {
+        if (IsConnectivityChange(result.status())) {
+          ++change_count;
+        }
+      });
+
+  SetNetworkStatus(NetworkStatus::Available);
+  // Same status shouldn't trigger a callback.
+  EXPECT_EQ(change_count, 0);
+
+  tester->KeepPollingGrpcQueue();
+  SetNetworkStatus(NetworkStatus::AvailableViaCellular);
+  EXPECT_EQ(change_count, 1);
+}
+
+TEST_F(GrpcConnectionTest, ConnectivityChangeWithSeveralActiveCalls) {
+  int changes_count = 0;
+
+  auto foo = tester->CreateStreamingReader();
+  foo->Start([&](const StatusOr<std::vector<grpc::ByteBuffer>>&) {
+    foo.reset();
+    ++changes_count;
+  });
+
+  auto bar = tester->CreateStreamingReader();
+  bar->Start([&](const StatusOr<std::vector<grpc::ByteBuffer>>&) {
+    bar.reset();
+    ++changes_count;
+  });
+
+  auto baz = tester->CreateStreamingReader();
+  baz->Start([&](const StatusOr<std::vector<grpc::ByteBuffer>>&) {
+    baz.reset();
+    ++changes_count;
+  });
+
+  tester->KeepPollingGrpcQueue();
+  EXPECT_NO_THROW(SetNetworkStatus(NetworkStatus::Unavailable));
```

Comment:
Is the `expect_no_throw` useful here? If yes, is it useful when setting the network status above (lines 137, 132, 118, 113)?

Reply:
The idea is to test this loop:
```cpp
for (GrpcCallInterface *call : calls) {
          // This will trigger the observers.
          call->FinishWithError(Status{FirestoreErrorCode::Unavailable,
                                       "Network connectivity changed"});
        }
```
and make sure that it can handle calls unregistering/deleting themselves during iteration. This check seems less interesting for the other cases where there is just a single call. Added a comment.